atsc/src/header.rs

```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-use bincode::{Decode, Encode};
+use log::{debug, trace};
+use std::panic;
 
-#[derive(Encode, Decode, Debug, Clone)]
+/*  The current file version.
+    On file read, compressors check the version and uncompress accordingly to that (of fail)
+*/
+const CURRENT_VERSION: u32 = 1;
+#[derive(Debug, Clone)]
 pub struct CompressorHeader {
     initial_segment: [u8; 4],
-    // We should go unsigned
-    frame_count: i16,
+    pub version: u32,
+    frame_count: u8,
+}
+
+fn verify_header_versions(version: u32) {
+    let current_version = CURRENT_VERSION;
+    trace!("Versions: c:{} h:{}", current_version, version);
+    match current_version.cmp(&version) {
+        std::cmp::Ordering::Less => panic!(
+            "Can't decompress! File is version ({}) is higher than compressor version ({})!",
+            version, current_version
+        ),
+        std::cmp::Ordering::Equal | std::cmp::Ordering::Greater => {
+            debug!("File version: {}", version)
+        }
+    }
 }
 
 impl CompressorHeader {
     pub fn new() -> Self {
         CompressorHeader {
             initial_segment: *b"BRRO",
-            // We have to limit the bytes of the header
+            version: CURRENT_VERSION,
             frame_count: 0,
         }
     }
 
     pub fn add_frame(&mut self) {
         self.frame_count += 1;
     }
+
+    pub fn get_frame_count(&mut self) -> u8 {
+        self.frame_count
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        // Converts header to a byte vector, little endian
+        let mut vec = Vec::new();
+        // Add initial_segment
+        vec.extend_from_slice(&self.initial_segment);
+        // Add version (u32 as 4 bytes)
+        vec.extend_from_slice(&self.version.to_le_bytes());
+        // Add frame_count
+        vec.push(self.frame_count);
+
+        vec
+    }
+
+    pub fn from_bytes(data: [u8; 9]) -> Self {
+        // Extract initial_segment
+        let initial_segment = [data[0], data[1], data[2], data[3]];
+        // Extract version (u32 from 4 bytes)
+        let version = u32::from_le_bytes([data[4], data[5], data[6], data[7]]);
```

Comment:
Since the version will always be the first part of the header to be parsed, I think the version should be the first 4 bytes starting at index 0 rather than starting at index 4.